File: backend-go/internal/notes/cursor.go
package notes

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"time"
)

// cursor encodes the stable sort key (updated_at, id) used to page
// through the notes feed, per /docs/adr/0007-cursor-pagination.md. It is
// opaque to clients; only this package constructs or interprets it.
type cursor struct {
	UpdatedAt string `json:"u"`
	ID        string `json:"i"`
}

func encodeCursor(updatedAt time.Time, id string) string {
	c := cursor{UpdatedAt: updatedAt.UTC().Format(time.RFC3339Nano), ID: id}
	b, _ := json.Marshal(c)
	return base64.RawURLEncoding.EncodeToString(b)
}

func decodeCursor(s string) (cursor, error) {
	b, err := base64.RawURLEncoding.DecodeString(s)
	if err != nil {
		return cursor{}, fmt.Errorf("invalid cursor: %w", err)
	}
	var c cursor
	if err := json.Unmarshal(b, &c); err != nil {
		return cursor{}, fmt.Errorf("invalid cursor: %w", err)
	}
	if c.UpdatedAt == "" || c.ID == "" {
		return cursor{}, fmt.Errorf("invalid cursor")
	}
	return c, nil
}

File: backend-rust/src/notes/cursor.rs
//! The stable sort key (`updated_at`, `id`) used to page through the notes
//! feed, per /docs/adr/0007-cursor-pagination.md. Opaque to clients; only
//! this module constructs or interprets it.

use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine as _};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Cursor {
    #[serde(rename = "u")]
    pub updated_at: String,
    #[serde(rename = "i")]
    pub id: String,
}

pub fn encode_cursor(updated_at: &str, id: &str) -> String {
    let c = Cursor {
        updated_at: updated_at.to_string(),
        id: id.to_string(),
    };
    let bytes = serde_json::to_vec(&c).expect("cursor serializes");
    URL_SAFE_NO_PAD.encode(bytes)
}

pub fn decode_cursor(s: &str) -> Result<Cursor, String> {
    let bytes = URL_SAFE_NO_PAD
        .decode(s)
        .map_err(|e| format!("invalid cursor: {e}"))?;
    let c: Cursor = serde_json::from_slice(&bytes).map_err(|e| format!("invalid cursor: {e}"))?;
    if c.updated_at.is_empty() || c.id.is_empty() {
        return Err("invalid cursor".to_string());
    }
    Ok(c)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cursor_round_trip() {
        let updated_at = "2026-07-29T12:00:00Z";
        let encoded = encode_cursor(updated_at, "note-123");

        let got = decode_cursor(&encoded).expect("decode");
        assert_eq!(got.id, "note-123");
        assert_eq!(got.updated_at, updated_at);
    }

    #[test]
    fn decode_cursor_rejects_garbage() {
        assert!(decode_cursor("not-valid-base64!!").is_err());
        // base64("{}"), missing fields
        assert!(decode_cursor("e30").is_err());
    }
}
